File: 199-binary-tree-right-side-view/solution.py
from typing import Optional
from collections import deque


class TreeNode:
  def __init__(self, val=0, left: Optional['TreeNode'] = None, right: Optional['TreeNode'] = None):
    self.val = val
    self.left = left
    self.right = right


class Solution:
  def rightSideView(self, root: Optional[TreeNode]) -> list[int]:
    result = []
    work_deque: deque[tuple[int, Optional[TreeNode]]] = deque([(0, root)])
    while work_deque:
      level, node = work_deque.pop()
      if node is None:
        continue

      work_deque.extend([(level + 1, node.left), (level + 1, node.right)])

      if len(result) == level:
        result.append(node.val)

    return result

File: 199-binary-tree-right-side-view/solution.cpp
#include <vector>
#include <deque>
#include <tuple>

struct TreeNode
{
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
  std::vector<int> rightSideView(TreeNode *root)
  {
    std::vector<int> result;
    std::deque<std::tuple<int, const TreeNode *>> work_deque{{0, root}};
    while (!work_deque.empty())
    {
      const auto &level_node = work_deque.back();
      work_deque.pop_back();

      const int level = std::get<0>(level_node);
      const TreeNode *node = std::get<1>(level_node);
      if (node == nullptr)
      {
        continue;
      }

      work_deque.emplace_back(level + 1, node->left);
      work_deque.emplace_back(level + 1, node->right);

      if (result.size() == level)
      {
        result.emplace_back(node->val);
      }
    }

    return result;
  }
};